### PR TITLE
Add authentication

### PR DESCRIPTION
**Pull request by lcbm**
**Describe what this PR introduces:** 
(if necessary, link currently open issues with [special keywords](https://help.github.com/en/articles/closing-issues-using-keywords))

- Add user authentication to the available operations. 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bug fix
- [X] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce any breaking changes?** (check one)

- [X] Yes
- [ ] No

If yes, please describe the impact and migration path for existing applications:
API consumers must send the [User Access Token (a.k.a. Authorization Key)](https://mainflux.readthedocs.io/en/latest/provisioning/#obtaining-an-authorization-key) in the `Authorization` Header, in order to successfully post and retrieve data.

**The PR fulfills these requirements:**

- [X] Related issues are referenced in the PR (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing
- [ ] New/updated tests are included

**Testing environment:**

- Operating System/Platform: Linux 5.5.13-arch2-1 x86_64 :computer: 
- Go version: go1.14.1 :rocket: 

If you have relevant logs, error output and etc, please attach them.

**Other information:**

Although this PR can be further improved, most of the changes required code code structure changes that are being addressed in pr #15. Applying these changes would imply in quite some rework. For this reason, sticking to the standard of the previous commits seemed a better approach -- so that when pr #15 is integrated, refactoring will be less overwhelming. 

Also, this PR is related to task [BLI-75](https://jira.cesar.org.br/browse/BLI-75) :confetti_ball:

### CODE REVIEW

File: pkg/network/cloud/requests.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package cloud
+
+// Request ...
```
Comment (lcbm) [inline pkg/network/cloud/requests.go]: Remove unnecessary comment :eyes:

File: pkg/interactor/cloud.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package interactor
+
+import (
+	"errors"
+
+	"github.com/CESARBR/knot-cloud-storage/pkg/network/cloud"
+)
+
```
Comment (lcbm) [inline pkg/interactor/cloud.go]: Add comment for exported error, bruh :roll_eyes:

File: pkg/interactor/cloud.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// CloudInteractor ...
+type CloudInteractor struct {
+	cloud cloud.Cloud
+}
+
+// NewCloudInteractor ...
+func NewCloudInteractor(cloud cloud.Cloud) CloudInteractor {
+	return CloudInteractor{cloud}
+}
```
Comment (lcbm) [inline pkg/interactor/cloud.go]: Constructors must return a ptrrrrr, jesus :pray:

File: pkg/interactor/cloud.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+// Authenticate ...
+func (ci *CloudInteractor) Authenticate(token string) error {
+	token, err := validateToken(token)
```
Comment (lcbm) [inline pkg/interactor/cloud.go]: too many `token` -- it's redundant :eyes: 
Also, bad practiceeee :sob: 
```suggestion
	t, err := validate(token)
```

File: pkg/interactor/cloud.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+	err = ci.cloud.Authenticate(token)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateToken(token string) (string, error) {
```
Comment (lcbm) [inline pkg/interactor/cloud.go]: Rename this as the name is redundant. :curly_loop: 
```suggestion
func validate(token string) (string, error) {
```

**Review by lcbm — COMMENTED** [review]
Self-reviewing is awesome :rocket: 
Git good, @lcbm :woman_facepalming: 

Obs.: wow, I cant request changes to myself. Consider it an order tho :angry: 
Obs. 2: quarantine is making me crazy.

File: pkg/interactor/cloud.go (hunk 3 above)
Comment (netoax) [reply pkg/interactor/cloud.go]: Take it easy, man

File: cmd/main.go [hunk 6]
```diff
@@ -21,10 +22,13 @@ func main() {
...
 	}
-
 	dataStore := data.NewDataStore(database)
 	dataInteractor := interactor.NewDataInteractor(dataStore)
-	dataController := controllers.NewDataController(dataInteractor)
+
+	cloud := cloud.NewCloud(config.Cloud.Host, config.Cloud.Port)
+	cloudInteractor := interactor.NewCloudInteractor(cloud)
+
+	dataController := controllers.NewDataController(dataInteractor, cloudInteractor)
```
Comment (lcbm) [inline cmd/main.go]: As discussed with @joaoaneto, instead of creating another interactor (which should be associated with a use-case, not to mention other architectural reasons), the cloud should be used the the `dataInteractor` :grimacing: 

Change this (and apply all other necessary modifications for this to work) :point_down: 
```suggestion
	dataStore := data.NewDataStore(database)
	cloud := cloud.NewCloud(config.Cloud.Host, config.Cloud.Port)
  dataInteractor := interactor.NewDataInteractor(dataStore, cloud)
	dataController := controllers.NewDataController(dataInteractor)
```

File: pkg/network/cloud/requests.go (hunk 1 above)
Comment (lcbm) [reply pkg/network/cloud/requests.go]: Actually fixed it instead :smile: 
Even if not exported, it's always helpful to have comments.

File: pkg/interactor/cloud.go (hunk 2 above)
Comment (lcbm) [reply pkg/interactor/cloud.go]: This actually moved elsewhere :thinking:

File: pkg/interactor/cloud.go (hunk 4 above)
Comment (lcbm) [reply pkg/interactor/cloud.go]: There was no need to return anything else other than the error itself. Done!

File: pkg/interactor/cloud.go (hunk 5 above)
Comment (lcbm) [reply pkg/interactor/cloud.go]: Done!

File: cmd/main.go (hunk 6 above)
Comment (lcbm) [reply cmd/main.go]: @joaoaneto done!

File: pkg/interactor/cloud.go (hunk 3 above)
Comment (lcbm) [reply pkg/interactor/cloud.go]: This interactor was removed :grimacing:

File: internal/config/default.yaml [hunk 7]
```diff
@@ -5,5 +5,9 @@ MongoDB:
   host: localhost
   name: things_db
 
+Cloud:
+  host: localhost
+  port: 8182
```
Comment (esdrasjnr) [inline internal/config/default.yaml]: Is that the correct port for babeltower service? Also suggest to use the service hostname (I think it's `bt`), what do you think?

**Review by esdrasjnr — CHANGES_REQUESTED** [review]
Excellent patch! Just made some minor questions :100:

File: internal/config/default.yaml (hunk 7 above)
Comment (lcbm) [reply internal/config/default.yaml]: Opsie, nice observation! Updated it :smile: 
Also, added `development` configurations for the cloud.

About changing `localhost` to `bt`, I'm not sure. For `bt` to work, the user would have to edit the DNS file so that `bt` points to `localhost`. When we deploy with `knot-cloud` and `docker swarm`, that's a different scenario, and we explicitly instruct the user to edit their `DNS` config file. 

What do you think @joaoaneto?

File: pkg/interactor/authenticator.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+import (
+	"errors"
+)
+
+// ErrTokenEmpty is returned for empty access tokens.
+var ErrTokenEmpty = errors.New("no access token provided")
+
+// Authenticate verifies if the access token is valid.
+func (d *DataInteractor) Authenticate(token string) error {
```
Comment (netoax) [inline pkg/interactor/authenticator.go]: According to what we discussed before, is it still necessary?

File: pkg/controllers/data.go [hunk 9]
```diff
@@ -45,38 +48,44 @@ func (d *DataController) GetAll(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	things, err := d.DataInteractor.GetAll(order, skip, take, startDate, finishDate)
+	token := r.Header.Get("Authorization")
```
Comment (netoax) [inline pkg/controllers/data.go]: In order to keep compatible with the [previous API](https://github.com/CESARBR/knot-cloud-storage#header-parameters-1), I think the header could be `auth_token`. What do you think?

File: pkg/interactor/authenticator.go (hunk 8 above)
Comment (netoax) [reply pkg/interactor/authenticator.go]: I got your intention after reviewing the [other](https://github.com/CESARBR/knot-cloud-storage/pull/18/commits/2c6603c1b9148b94d3925cec5d8cf457a9aa9062) commit. However, this method is still placed as it was a use case operation. Considering our data use cases are concentrated at the `interactors/data.go` file, I think it's fine moving this method there.

File: internal/config/default.yaml [hunk 10]
```diff
@@ -5,5 +5,9 @@ MongoDB:
   host: localhost
   name: things_db
 
+Cloud:
```
Comment (netoax) [inline internal/config/default.yaml]: Why did you choose `cloud` instead of `users`?

File: pkg/network/cloud/cloud.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
+package cloud
+
+import (
+	"fmt"
+
+	"github.com/CESARBR/knot-cloud-storage/pkg/logging"
+)
+
+// Cloud is an interface that defines the cloud's use cases' operations.
+type Cloud interface {
```
Comment (netoax) [inline pkg/network/cloud/cloud.go]: I'd suggest you rename to `CloudService` or even change the point of view and see this interface as an `Authenticator`, maybe? IMO, it's better than having the same name as the  `cloud` structure.

File: pkg/network/cloud/transport.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
+package cloud
```
Comment (netoax) [inline pkg/network/cloud/transport.go]: Is there a specific reason for this file being named `transport`?

**Review by netoax — CHANGES_REQUESTED** [review]

File: internal/config/default.yaml (hunk 7 above)
Comment (netoax) [reply internal/config/default.yaml]: I think `default` configuration should be neutral apart from deploying specificities. So, `localhost` and `80` fits well.

File: pkg/interactor/authenticator.go (hunk 8 above)
Comment (lcbm) [reply pkg/interactor/authenticator.go]: Considering the sprint deadline and that we won't be refactoring this code anytime soon... I'm moving the method there. However, I don't think it's a good practice. This could become a file with many hundreds of lines of code. It's seems a better approach to separate each use case in a separate file.

File: internal/config/default.yaml (hunk 7 above)
Comment (lcbm) [reply internal/config/default.yaml]: Done :smile:

File: pkg/controllers/data.go (hunk 9 above)
Comment (lcbm) [reply pkg/controllers/data.go]: Agreed! Done :smile:

File: internal/config/default.yaml (hunk 10 above)
Comment (lcbm) [reply internal/config/default.yaml]: Makes sense! :smile: 
Done!

File: pkg/network/cloud/cloud.go (hunk 11 above)
Comment (lcbm) [reply pkg/network/cloud/cloud.go]: IMO, adding `Service` is redundant. If we need to specify what it is, may we aren't being clear enough with context and package structure. I Agree that `Cloud` isn't good and I'll make some changes, let me know what you think.
>Heres an [example](https://github.com/mainflux/mainflux/blob/master/authn/service.go)

File: pkg/network/cloud/transport.go (hunk 12 above)
Comment (lcbm) [reply pkg/network/cloud/transport.go]: It handles operations on the transport layer. Its a well-known convention in package oriented design in Golang.

File: pkg/network/cloud/cloud.go (hunk 11 above)
Comment (lcbm) [reply pkg/network/cloud/cloud.go]: Done, let me know what you think.

File: pkg/interactor/authenticator.go (hunk 8 above)
Comment (netoax) [reply pkg/interactor/authenticator.go]: Yes, I definitely agree that each use case should be in a separate file and that's the approach we usually follow. However, considering our reality, moving continues being the best solution to avoid the code become worse.

Comment (netoax) [reply pkg/interactor/authenticator.go]: Thanks!

File: pkg/network/cloud/cloud.go (hunk 11 above)
Comment (netoax) [reply pkg/network/cloud/cloud.go]: I'd say `Service` can be redundant if we look to it as the example you mentioned does. In that case, it represents the interface to the service's domain capabilities (aka use cases or interactions). Instead, we want to create an interface to the **external** `users` service and that's why in clean architecture we usually place these components in the `network` or `data` directory (depending on what it provides).

What do you think about `Authenticator`? It seems to fit well in terms of interface naming best practices and we gain in flexibility to choose the struct name.

File: pkg/users/service.go [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+// Service specifies an API to the domain service.
+type Service interface {
+	// Authenticate validates the given token. If it is valid,
+	// nil error value is returned. If the token in invalid,
+	// a non-nil error value is returned.
+	Authenticate(token string) error
+}
+
+var _ Service = (*service)(nil)
```
Comment (netoax) [inline pkg/users/service.go]: I found it hard to understand this: `var _ Service = (*service)(nil)`. Is it necessary?

It seems to check statically (compiler time) if the interface is really implemented. Even finding good explanations [here](https://stackoverflow.com/a/13194635) and [here](https://www.reddit.com/r/golang/comments/8zw964/why_would_you_declare_var_with_blank_identify/), there's no consensus about if it's a good or bad practice.

As a matter of personal preference, I wouldn't go with that.

File: pkg/users/service.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package users
+
+import (
+	"fmt"
+
+	"github.com/CESARBR/knot-cloud-storage/pkg/logging"
+)
+
+// Service specifies an API to the domain service.
+type Service interface {
```
Comment (netoax) [inline pkg/users/service.go]: Shouldn't this change be in another commit?

File: pkg/users/service.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+	// Authenticate validates the given token. If it is valid,
+	// nil error value is returned. If the token in invalid,
+	// a non-nil error value is returned.
+	Authenticate(token string) error
```
Comment (netoax) [inline pkg/users/service.go:15]: Don't you think this method is self-explained? I think the comment can be removed.

**Review by netoax — CHANGES_REQUESTED** [review]
Good improvements, @lcbm. Thanks!
Just some remarks.

File: pkg/users/service.go (hunk 14 above)
Comment (lcbm) [reply pkg/users/service.go]: I must've done a bad `-i` rebase... lol! done :smile:

File: pkg/network/cloud/cloud.go (hunk 11 above)
Comment (lcbm) [reply pkg/network/cloud/cloud.go]: `Authenticator` seems good, done :smile:

File: pkg/users/service.go (hunk 13 above)
Comment (lcbm) [reply pkg/users/service.go]: Done :smile:

File: pkg/users/service.go (hunk 15 above)
Comment (lcbm) [reply pkg/users/service.go:15]: Done :smile:

**Review by netoax — APPROVED** [review]
LGTM!
